File: typescipt/src/language/types/functionType.ts


export class FunctionType extends TypeWithMembers {
   public string Type
   public Function Function

   constructor(type: string, function: Function) {
     Type = type;
     Function = function;
   }

   protected equals(other: FunctionType): boolean {
     return Type == other.Type;
   }

   public override equals(obj: object): boolean {
     if (ReferenceEquals(null, obj)) return false;
     if (ReferenceEquals(this, obj)) return true;
     if (obj.GetType() != GetType()) return false;
     return Equals((FunctionType)obj);
   }

   public override getHashCode(): number {
     return Type != null ? Type.GetHashCode() : 0;
   }

   public override toString(): string {
     return Type;
   }

   public override memberType(name: string, context: IValidationContext): VariableType {
     return name switch {
       Function.ParameterName => FunctionParametersType(context),
       Function.resultsName => FunctionResultsType(context),
       _ => null
     };
   }

   private functionParametersType(context: IValidationContext): FunctionParametersType {
     let complexType = context.RootNodes.GetFunction(Type)?.GetParametersType(context);
     return new FunctionParametersType(Type, complexType);
   }

   private functionResultsType(context: IValidationContext): FunctionResultsType {
     let complexType = context.RootNodes.GetFunction(Type)?.GetResultsType(context);
     return new FunctionResultsType(Type, complexType);
   }
}

File: src/Lexy.Compiler/Language/VariableTypes/FunctionType.cs
using Lexy.Compiler.Language.Functions;

namespace Lexy.Compiler.Language.VariableTypes;

public class FunctionType : TypeWithMembers
{
    public string Type { get; }
    public Function Function { get; }

    public FunctionType(string type, Function function)
    {
        Type = type;
        Function = function;
    }

    protected bool Equals(FunctionType other)
    {
        return Type == other.Type;
    }

    public override bool Equals(object obj)
    {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != GetType()) return false;
        return Equals((FunctionType)obj);
    }

    public override int GetHashCode()
    {
        return Type != null ? Type.GetHashCode() : 0;
    }

    public override string ToString()
    {
        return Type;
    }

    public override VariableType MemberType(string name, IComponentNodeList componentNodes)
    {
        return name switch
        {
            Function.ParameterName => FunctionParametersType(componentNodes),
            Function.ResultsName => FunctionResultsType(componentNodes),
            _ => null
        };
    }

    private ComplexType FunctionParametersType(IComponentNodeList componentNodes)
    {
        return componentNodes.GetFunction(Type)?.GetParametersType();
    }

    private ComplexType FunctionResultsType(IComponentNodeList componentNodes)
    {
        return componentNodes.GetFunction(Type)?.GetResultsType() as ComplexType;
    }
}